From: "Liam R. Howlett" <Liam.Howlett@oracle.com>
Subject: mm/mmap: reorder validate_mm_mt() checks
Date: Tue, 28 Jun 2022 15:08:19 -0400

Allows for detecting broken trees before attempting to iterate over the tree.

Link: https://lkml.kernel.org/r/20220629153127.p2k2cqkc7qy7f3vw@revolver
Fixes: de4583ed958b (mm: start tracking VMAs with maple tree)
Signed-off-by: Liam R. Howlett <Liam.Howlett@oracle.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/mmap.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/mm/mmap.c~mm-start-tracking-vmas-with-maple-tree-fix
+++ a/mm/mmap.c
@@ -392,6 +392,8 @@ static void validate_mm_mt(struct mm_str
 	struct vm_area_struct *vma_mt, *vma = mm->mmap;
 
 	MA_STATE(mas, mt, 0, 0);
+
+	mt_validate(&mm->mm_mt);
 	mas_for_each(&mas, vma_mt, ULONG_MAX) {
 		if (xa_is_zero(vma_mt))
 			continue;
@@ -440,7 +442,6 @@ static void validate_mm_mt(struct mm_str
 
 	}
 	VM_BUG_ON(vma);
-	mt_validate(&mm->mm_mt);
 }
 #else
 #define validate_mm_mt(root) do { } while (0)
_
